merino-suggest/src/keyword_filter.rs

```diff
@@ -0,0 +1,223 @@
+//! A suggestion provider that filters suggestions from a subprovider.
+
+use std::collections::HashMap;
+
+use crate::{SetupError, SuggestionProvider, SuggestionResponse};
+use anyhow::anyhow;
+use async_trait::async_trait;
+use cadence::{Counted, StatsdClient};
+use regex::Regex;
+
+/// A combinator provider that filters the results from the wrapped provider
+/// using a blocklist from the settings.
+pub struct KeywordFilterProvider {
+    /// A map linking the filter identifiers to their compiled regexes.
+    blocklist: HashMap<String, regex::Regex>,
+
+    /// The provider to pull suggestions from.
+    inner: Box<dyn SuggestionProvider>,
+}
+
+impl KeywordFilterProvider {
+    /// Construct a new, boxed filter provider.
+    pub fn new_boxed(
+        blocklist: HashMap<String, String>,
+        inner: Box<dyn SuggestionProvider>,
+    ) -> Result<Box<Self>, SetupError> {
+        // Compile the provided blocklist regexes just once.
+        let mut compiled_blocklist: HashMap<String, Regex> = HashMap::new();
+
+        for filter in &blocklist {
+            if let Ok(r) = Regex::new(filter.1) {
+                compiled_blocklist.insert(filter.0.to_string(), r);
```

Comment:
How would you feel about naming the parts of the filter, instead of using `filter.0` and `filter.1`?

```suggestion
        for (filter_id, filter_regex) in &blocklist {
            if let Ok(r) = Regex::new(filter_regex) {
                compiled_blocklist.insert(filter_id.to_string(), r);
```